### ISSUE TITLE
Entity Identity instead of state used for assertIsEqualToComparingFieldByFieldRecursively

### ISSUE THREAD

**Issue by mjeffrey** [joel-costigliola/assertj-core#886]
#### Summary

When there is an equals() method defined then Objects::assertIsEqualToComparingFieldByFieldRecursively does not compare the remainder of the fields on the Object.

I made some suggestions for fixes below. Let me know what you think, I can create a pull request with what I did.

In JPA/Hibernate the identity of entity objects (really just POJOs) is important. So you usually need an equals/hashcode to work effectively with them (Object identity can be used as well but it has some drawbacks). Usually this identity is not based on all fields of the class so there are usually fields which don't make up part of the equals/hashcode methods and form the entity's state. 
If you use Objects::assertIsEqualToComparingFieldByFieldRecursively then the Object's equals() method is used to make the comparison and effectively ignores the fields that are not part of the equals() method.

The issue is in DeepDifference::determineDifferences in the second continue below. Once the equals is check we skip checking the fields.
```java
if (hasCustomEquals(key1.getClass())) {
  if (!(key1.equals(key2))) 
    differences.add(new Difference(currentPath, key1, key2));
    continue;
  }
  continue;
}
```
One option I used to fix this is to make the second "continue" conditional on a set of primitive and other classes. Any of these classes would behave as before and only check equals() but any other classes would do a field by field comparison in addition to the equals() check. It is important to exclude some classes since an "int" is decomposed to another "int" and with auto unboxing there is infinite decomposition. The downside with this approach is that it is possible we could introduce infinite recursion (restrict path length?) in some cases. And maybe there are other classes that need to be excluded as well

```java
if (hasCustomEquals(key1.getClass())) {
  if (!(key1.equals(key2))) 
    differences.add(new Difference(currentPath, key1, key2));
    continue;
  }
  if ( compareClassUsingOnlyEquals(key1)){
    continue;
  }
}
```

Another option I tried was to exclude all java/javax classes from the decomposition and just rely on their equals() method. This is probably what users expect (people wouldn't expect that we examine the internal state of each jdk object). However it could be that there are other packages that should also be excluded (in this case I suppose the user can already exclude properties).

Example 1: This test passes when it should fail. The test can be made to fail by just removing the second continue (but that causes infinite decomposition in the second example dues to the lineNums being equal and being recursively decomposed to fields).

```java
  class EqualAddress extends Address{
    @Override
    public boolean equals(Object obj) {
      return true;
    }
  }
  @Test
  public void should_treat_state_differences_as_not_equals() {
    Person actual = new Person();
    actual.name = "F Dagg";
    actual.home.address = new EqualAddress();
    actual.home.address.number = 1;

    Person other = new Person();
    other.name = "F Dagg";
    other.home.address = new EqualAddress();
    other.home.address.number = 20000;

    objects.assertIsEqualToComparingFieldByFieldRecursively(someInfo(), actual, other, noFieldComparators(),
            defaultTypeComparators());
  }
```

#### Example 2: Order entity with a set of Order Lines. This test passes when it should fail.

```java
  @Test
  public void should_treat_different_state_as_difference() {
    Order actual = new Order(1);
    actual.addOrderLine(new OrderLine("burger", 2 ));

    Order other = new Order(1);
    other.addOrderLine(new OrderLine("fries", 1 ));

    objects.assertIsEqualToComparingFieldByFieldRecursively(someInfo(), actual, other, noFieldComparators(),
            defaultTypeComparators());
  }

  public static class Order{
    private final int orderId;
    private Set<OrderLine> orderLines = new HashSet<>();
    public String description;

    public Order(int orderId) {
      this.orderId = orderId;
    }

    public void addOrderLine(OrderLine orderLine){
      orderLine.order = this;
      orderLine.lineNum = orderLines.size() + 1;
      orderLines.add(orderLine);
    }

    @Override // generated by IDE for orderId
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      Order order = (Order) o;

      return orderId == order.orderId;
    }

    @Override // generated by IDE for orderId
    public int hashCode() {
      return orderId;
    }
  }
  public static class OrderLine {
    public Order order;
    public int lineNum;
    public String productCode;
    public int quantity;

    public OrderLine(String productCode, int quantity) {
      this.productCode = productCode;
      this.quantity = quantity;
    }

    @Override // generated by IDE for order and lineNum
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      OrderLine orderLine = (OrderLine) o;

      if (lineNum != orderLine.lineNum) return false;
      return order.equals(orderLine.order);
    }

    @Override // generated by IDE for order and lineNum
    public int hashCode() {
      int result = order.hashCode();
      result = 31 * result + lineNum;
      return result;
    }
  }
```
#### Java 8 specific ?

* NO  : create Pull Request from the `2.x` branch

**Comment by joel-costigliola** [comment]
We won't change the current behavior of `assertIsEqualToComparingFieldByFieldRecursively` because it is a breaking change and if you have overridden equals it should be trusted to assess equality.

What we could do is provide a way comparing graph objects ignoring any custom equals except for basic types (JDK ones), would that work for you ?

**Comment by mjeffrey** [comment]
Yes, essentially that is what I was proposing with the code snippet below where I made the second continue conditional. 
This avoids strange cases where you have an equals() on Person based on the social security number but all other data is different (and they get treated as equal). In tests I think we are more interested in state than identity. 

```java
if (hasCustomEquals(key1.getClass())) {
  if (!(key1.equals(key2))) 
    differences.add(new Difference(currentPath, key1, key2));
    continue;
  }
  if ( compareClassUsingOnlyEquals(key1)){
    continue;
  }
}
```

The method compareClassUsingOnlyEquals() would return true for java primitive wrappers (and a few other things like Date/Timestamp) or maybe even better, for all classes names starting with "java." or "javax.").

**Comment by PascalSchumacher** [comment]
I guess this issue is a duplicate or at least a close "relative" of https://github.com/joel-costigliola/assertj-core/issues/646

**Comment by joel-costigliola** [comment]
Agreed.
Closed as duplicate of #646.
